### PR TITLE
Add Scheme, Host and Port to base cache key

### PR DESCRIPTION
**Pull request by JunTaoLuo**
#86

### CODE REVIEW

File: src/Microsoft.AspNetCore.ResponseCaching/Internal/StringBuilderExtensions.cs [hunk 1]
```diff
@@ -10,10 +9,13 @@ internal static class StringBuilderExtensions
     {
         internal static StringBuilder AppendUpperInvariant(this StringBuilder builder, string value)
         {
-            builder.EnsureCapacity(builder.Length + value.Length);
-            for (var i = 0; i < value.Length; i++)
+            if (value != null)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.ResponseCaching/Internal/StringBuilderExtensions.cs]: `string.IsNullOrEmpty` seems better

**Comment by natemcmaster** [comment]
@JunTaoLuo a bunch of tests are failing now with NRE.

```
System.NullReferenceException : Object reference not set to an instance of an object.
Error stack trace:
   at Microsoft.AspNetCore.ResponseCaching.Internal.StringBuilderExtensions.AppendUpperInvariant(StringBuilder builder, String value) in C:\projects\responsecaching\src\Microsoft.AspNetCore.ResponseCaching\Internal\StringBuilderExtensions.cs:line 12
   at Microsoft.AspNetCore.ResponseCaching.Internal.ResponseCachingKeyProvider.CreateBaseKey(ResponseCachingContext context) in C:\projects\responsecaching\src\Microsoft.AspNetCore.ResponseCaching\Internal\ResponseCachingKeyProvider.cs:line 51
   at Microsoft.AspNetCore.ResponseCaching.ResponseCachingMiddleware.OnFinalizeCacheHeaders(ResponseCachingContext context) in C:\projects\responsecaching\src\Microsoft.AspNetCore.ResponseCaching\ResponseCachingMiddleware.cs:line 260
   at Microsoft.AspNetCore.ResponseCaching.ResponseCachingMiddleware.FinalizeCacheHeadersAsync(ResponseCachingContext context) in C:\projects\responsecaching\src\Microsoft.AspNetCore.ResponseCaching\ResponseCachingMiddleware.cs:line 331
   at Microsoft.AspNetCore.ResponseCaching.Tests.ResponseCachingMiddlewareTests.<FinalizeCacheHeadersAsync_UpdateCachedVaryByRules_IfNotEquivalentToPrevious>d__26.MoveNext() in C:\projects\responsecaching\test\Microsoft.AspNetCore.ResponseCaching.Tests\ResponseCachingMiddlewareTests.cs:line 550
--- End of stack trace from previous location where exception was thrown ---
--- End of stack trace from previous location where exception was thrown ---
--- End of stack trace from previous location where exception was thrown ---
```

File: src/Microsoft.AspNetCore.ResponseCaching/Internal/StringBuilderExtensions.cs [hunk 2]
```diff
@@ -10,10 +9,13 @@ internal static class StringBuilderExtensions
     {
         internal static StringBuilder AppendUpperInvariant(this StringBuilder builder, string value)
         {
-            builder.EnsureCapacity(builder.Length + value.Length);
-            for (var i = 0; i < value.Length; i++)
+            if (string.IsNullOrEmpty(value))
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.ResponseCaching/Internal/StringBuilderExtensions.cs]: Should this be inverted? NOT string.IsNullOrEmpty?

**Comment by JunTaoLuo** [comment]
Hmm, let me check. I was able to run the tests locally.

File: src/Microsoft.AspNetCore.ResponseCaching/Internal/StringBuilderExtensions.cs (hunk 2 above)
Comment (JunTaoLuo) [reply src/Microsoft.AspNetCore.ResponseCaching/Internal/StringBuilderExtensions.cs]: Yup, bad typo during fixup.

Comment (JunTaoLuo) [reply src/Microsoft.AspNetCore.ResponseCaching/Internal/StringBuilderExtensions.cs]: Pushed fix, ran the tests again to make sure.

Comment (natemcmaster) [reply src/Microsoft.AspNetCore.ResponseCaching/Internal/StringBuilderExtensions.cs]: Thanks!

**Pull request merged**